In [26]:
import pandas as pd
import arrow
from darts import TimeSeries
from datetime import datetime

In [97]:
data = pd.read_csv('busan_incheon_hadong_solarratio_temp_pre_pm_time2d_2013_2020.csv', index_col=0)

In [98]:
data

Busan_Solar_Ratio  Busan_Temp  Busan_Precipitation  Busan_PM10  \
0                    0.0        -2.3                  NaN        40.0   
1                    0.0        -2.4                  NaN        39.0   
2                    0.0        -2.5                  NaN        47.0   
3                    0.0        -2.0                  NaN        42.0   
4                    0.0        -2.3                  NaN        42.0   
...                  ...         ...                  ...         ...   
70123                0.0         0.6                  NaN        21.0   
70124                0.0         0.1                  NaN        20.0   
70125                0.0        -0.2                  NaN        12.0   
70126                0.0        -0.7                  NaN        20.0   
70127                0.0        -0.7                  NaN        24.0   

       Incheon_Solar_Ratio  Incheon_Temp  Incheon_Precipitation  Incheon_PM10  \
0                      0.0          -6.9                    NaN          43.0   
1                      0.0          -6.9                    NaN          36.0   
2                      0.0          -6.7                    NaN          30.0   
3                      0.0          -6.5                    NaN          30.0   
4                      0.0          -6.3                    NaN          28.0   
...                    ...           ...                    ...           ...   
70123                  0.0          -6.9                    NaN          26.0   
70124                  0.0          -7.2                    NaN          36.0   
70125                  0.0          -6.8                    NaN          37.0   
70126                  0.0          -6.3                    NaN          33.0   
70127                  0.0          -6.8                    NaN          20.0   

       Hadong_Solar_Ratio  Hadong_Temp  Hadong_Precipitation  Hadong_PM10  \
0                     0.0         -1.2                   NaN         39.0   
1                     0.0         -1.5                   NaN         16.0   
2                     0.0         -1.9                   NaN         17.0   
3                     0.0         -2.0                   NaN         15.0   
4                     0.0         -1.8                   NaN         17.0   
...                   ...          ...                   ...          ...   
70123                 0.0         -0.4                   NaN         17.0   
70124                 0.0         -0.9                   NaN         21.0   
70125                 0.0         -1.1                   NaN         19.0   
70126                 0.0         -0.9                   NaN         21.0   
70127                 0.0         -0.4                   NaN         22.0   

         Year_X    Year_Y    Time_X        Time_Y                       date  
0      1.000000  0.000000  0.965926  2.588190e-01  2013-01-01T01:00:00+09:00  
1      1.000000  0.000000  0.866025  5.000000e-01  2013-01-01T02:00:00+09:00  
2      1.000000  0.000000  0.707107  7.071068e-01  2013-01-01T03:00:00+09:00  
3      1.000000  0.000000  0.500000  8.660254e-01  2013-01-01T04:00:00+09:00  
4      1.000000  0.000000  0.258819  9.659258e-01  2013-01-01T05:00:00+09:00  
...         ...       ...       ...           ...                        ...  
70123  0.999852 -0.017213  0.500000 -8.660254e-01  2020-12-31T20:00:00+09:00  
70124  0.999852 -0.017213  0.707107 -7.071068e-01  2020-12-31T21:00:00+09:00  
70125  0.999852 -0.017213  0.866025 -5.000000e-01  2020-12-31T22:00:00+09:00  
70126  0.999852 -0.017213  0.965926 -2.588190e-01  2020-12-31T23:00:00+09:00  
70127  0.999852  0.017213  1.000000 -2.449294e-16  2021-01-01T00:00:00+09:00  

[70128 rows x 17 columns]

In [99]:
for i in data.index:
    data['date'][i] = arrow.get(data['date'][i]).format('YYYY-MM-DD HH:MM:SS')

<ipython-input-99-091dca82c1c3>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['date'][i] = arrow.get(data['date'][i]).format('YYYY-MM-DD HH:MM:SS')


In [100]:
data['date']

0        2013-01-01 01:01:00
1        2013-01-01 02:01:00
2        2013-01-01 03:01:00
3        2013-01-01 04:01:00
4        2013-01-01 05:01:00
                ...         
70123    2020-12-31 20:12:00
70124    2020-12-31 21:12:00
70125    2020-12-31 22:12:00
70126    2020-12-31 23:12:00
70127    2021-01-01 00:01:00
Name: date, Length: 70128, dtype: object

In [86]:
datetime.strptime(str(data['date'][0]), 'YYYYMMDDHH')

ValueError: time data '2013010101' does not match format 'YYYYMMDDHH'

In [64]:
data['date_idx'] = data.index

In [65]:
data

Busan_Solar_Ratio  Busan_Temp  Busan_Precipitation  Busan_PM10  \
0                    0.0        -2.3                  NaN        40.0   
1                    0.0        -2.4                  NaN        39.0   
2                    0.0        -2.5                  NaN        47.0   
3                    0.0        -2.0                  NaN        42.0   
4                    0.0        -2.3                  NaN        42.0   
...                  ...         ...                  ...         ...   
70123                0.0         0.6                  NaN        21.0   
70124                0.0         0.1                  NaN        20.0   
70125                0.0        -0.2                  NaN        12.0   
70126                0.0        -0.7                  NaN        20.0   
70127                0.0        -0.7                  NaN        24.0   

       Incheon_Solar_Ratio  Incheon_Temp  Incheon_Precipitation  Incheon_PM10  \
0                      0.0          -6.9                    NaN          43.0   
1                      0.0          -6.9                    NaN          36.0   
2                      0.0          -6.7                    NaN          30.0   
3                      0.0          -6.5                    NaN          30.0   
4                      0.0          -6.3                    NaN          28.0   
...                    ...           ...                    ...           ...   
70123                  0.0          -6.9                    NaN          26.0   
70124                  0.0          -7.2                    NaN          36.0   
70125                  0.0          -6.8                    NaN          37.0   
70126                  0.0          -6.3                    NaN          33.0   
70127                  0.0          -6.8                    NaN          20.0   

       Hadong_Solar_Ratio  Hadong_Temp  Hadong_Precipitation  Hadong_PM10  \
0                     0.0         -1.2                   NaN         39.0   
1                     0.0         -1.5                   NaN         16.0   
2                     0.0         -1.9                   NaN         17.0   
3                     0.0         -2.0                   NaN         15.0   
4                     0.0         -1.8                   NaN         17.0   
...                   ...          ...                   ...          ...   
70123                 0.0         -0.4                   NaN         17.0   
70124                 0.0         -0.9                   NaN         21.0   
70125                 0.0         -1.1                   NaN         19.0   
70126                 0.0         -0.9                   NaN         21.0   
70127                 0.0         -0.4                   NaN         22.0   

         Year_X    Year_Y    Time_X        Time_Y        date  date_idx  
0      1.000000  0.000000  0.965926  2.588190e-01  2013010101         0  
1      1.000000  0.000000  0.866025  5.000000e-01  2013010102         1  
2      1.000000  0.000000  0.707107  7.071068e-01  2013010103         2  
3      1.000000  0.000000  0.500000  8.660254e-01  2013010104         3  
4      1.000000  0.000000  0.258819  9.659258e-01  2013010105         4  
...         ...       ...       ...           ...         ...       ...  
70123  0.999852 -0.017213  0.500000 -8.660254e-01  2020123120     70123  
70124  0.999852 -0.017213  0.707107 -7.071068e-01  2020123121     70124  
70125  0.999852 -0.017213  0.866025 -5.000000e-01  2020123122     70125  
70126  0.999852 -0.017213  0.965926 -2.588190e-01  2020123123     70126  
70127  0.999852  0.017213  1.000000 -2.449294e-16  2021010100     70127  

[70128 rows x 18 columns]

In [69]:
series = TimeSeries.from_dataframe(data, 'date_idx', 'Busan_Solar_Ratio')

In [71]:
train, val = series[:int(len(series)*0.6)], series[int(len(series)*0.6):]

In [72]:
train

<TimeSeries (DataArray) (date_idx: 42076, component: 1, sample: 1)>
array([[[0.]],

       [[0.]],

       [[0.]],

       ...,

       [[0.]],

       [[0.]],

       [[0.]]])
Coordinates:
  * date_idx   (date_idx) int64 0 1 2 3 4 5 ... 42071 42072 42073 42074 42075
  * component  (component) object 'Busan_Solar_Ratio'
Dimensions without coordinates: sample

data['time_idx'] = [i for i in range(0, 70128)]
data['groups'] = 0


In [56]:
data = data.iloc[8760:]

In [57]:
data = data.fillna(0)


In [58]:
max_prediction_length = 6
max_encoder_length = 24
training_cutoff = data['time_idx'].max() - max_prediction_length

training = TimeSeriesDataSet(
    data[lambda x: x.time_idx <= training_cutoff],
    time_idx='time_idx',
    target='Busan_Solar_Ratio',
    group_ids=['groups'],
    min_encoder_length=max_encoder_length//2,
    max_encoder_length=max_encoder_length,
    min_prediction_length=1,
    max_prediction_length=max_prediction_length,
    time_varying_known_reals=['Busan_Temp', 'Busan_Precipitation', 'Busan_PM10',
       'Incheon_Temp', 'Incheon_Precipitation',
       'Incheon_PM10', 'Year_X', 'Year_Y', 'Time_X',
       'Time_Y', 'time_idx'], #'Hadong_Solar_Ratio', 'Hadong_Temp','Hadong_Precipitation', 'Hadong_PM10', 
    time_varying_unknown_reals=['Busan_Solar_Ratio', 'Incheon_Solar_Ratio']
    
)

NameError: name 'TimeSeriesDataSet' is not defined

In [ ]:
for i in training.to_dataloader():
    print(i)
    break

({'encoder_cat': tensor([], size=(64, 24, 0), dtype=torch.int64), 'encoder_cont': tensor([[[ 1.0000, -1.4135, -0.1253,  ..., -1.2460, -0.6364, -0.6395],
         [ 1.0000, -1.4970, -0.1253,  ..., -1.2459, -0.6364, -0.6395],
         [ 1.0000, -1.5925, -0.1253,  ..., -1.2458, -0.6364, -0.6395],
         ...,
         [ 1.0000, -0.9721, -0.1253,  ..., -1.2448,  0.1865,  1.2449],
         [ 1.0000, -1.0079, -0.1253,  ..., -1.2447, -0.4788, -0.2034],
         [ 1.0000, -1.2346, -0.1253,  ..., -1.2447, -0.6364, -0.6382]],

        [[ 1.0000,  1.4138, -0.1253,  ..., -0.4595, -0.7397, -0.6395],
         [ 1.0000,  1.4377, -0.1253,  ..., -0.4594, -0.4895, -0.6272],
         [ 1.0000,  1.4377, -0.1253,  ..., -0.4594, -0.3280, -0.4439],
         ...,
         [ 1.0000,  1.3065, -0.1253,  ..., -0.4583, -0.7478, -0.6395],
         [ 1.0000,  1.3065, -0.1253,  ..., -0.4583, -0.7478, -0.6395],
         [ 1.0000,  1.3422, -0.1253,  ..., -0.4582, -0.7478, -0.6395]],

        [[ 1.0000, -1.2465, -0.125